bpf/complexity-tests/61/bpf_host/7.txt

```diff
@@ -0,0 +1,33 @@
+-DHAVE_LARGE_INSN_LIMIT=1
```

Comment:
I don't think we've 100% gotten to the bottom of this yet. Regardless, this is where we are based on my current findings:

Firstly, this reproduction actually works on multiple Kernel versions, I was testing it on v1.15 and v6.2 locally.  I'm fairly confident that this isn't particular to a specific Kernel version (at least not any newer versions).

One key factor in reproducing this appears to be the version of Clang/LLVM being used.  [This](https://github.com/cilium/cilium/pull/30431/commits/49e85e0679dc46aca439b95997546d814e66e8fe) commit bumps the lvh images based on [this version update](https://github.com/cilium/little-vm-helper-images/pull/332/commits/43f8dbc302580db6027eafccd0c7d4a3b7b6a73c), which brings the CI Clang version to match what we use in Cilium Agent/Builder images.

Prior to this, I could not induce the failure on CI (locally, I was using the latest cilium-builder image to match what the Agent runs):
```
clang version 10.0.0 (https://github.com/llvm/llvm-project.git 0598a534371d5fd6debd129b1378b39b923b9787)
```

Now as for this file (7.txt),  through trial and error, I've identified that the necessary conditions for this to happen are:

```
      HAVE_LARGE_INSN_LIMIT=1
      ENABLE_IPV4=1
      ENABLE_HOST_FIREWALL=1
      ENABLE_MASQUERADE_IPV4=1
      ENABLE_NODEPORT=1
      ENABLE_IPV4_FRAGMENTS=1
```
 
This is actually not uncommon, and is in parity with  what 1.txt in directory sets.  So this file is essentially identical to 1.txt, with one key change, this omits:

`POLICY_VERDICT_NOTIFY`

Which corresponds with macroing out [this block of code ](https://github.com/cilium/cilium/blob/main/bpf/lib/policy_log.h#L86-L94) in the translation unit.  This is quite interesting because this is the _negative_ case which creates `send_policy_verdict_notify` as an inlined function that doesn't do anything.

However this seemingly innocuous difference bumps the complexity from 400k to over a million.

This is very counter intuitive, one possibility is that this code is creating some kind of side effect somewhere else in the preprocessor?   However, because this seems to depend on LLVM versions to happen, we may want to consider the possibility that this is a compiler bug.

One last note, disabling `POLICY_VERDICT_NOTIFY` is not actually representative of my local reproductions while running Cilium, in fact it seems like in practice this is always set `POLICY_VERDICT_NOTIFY=1`. This is just something I arrived at through some trial-and-error locally (the first set of values did come from my live cluster reproduction however).
I'm fairly confident there is some other, more realistic, combination of variables that I can distill from my local environment that would cause this same issue.

I think the key point of this is, the very non-linear relationship between some of these code changes and the number of instructions, coupled with the dependency on clang versions for reproduction, also suggests to me that something strange is going on with the compiler.

Reply:
So, ultimately, I think we need to dig into to this and understand the root issue here, as there may be other cases of this issue lurking.

@julianwiedmann's changes should dramatically lower the overall complexity of the host ingress path so in the meantime this seems like a acceptable way to unblock us for v1.15.

Reply:
Curiously enough, v1.15 CI still doesn't reproduce the problem with a config derived from https://github.com/cilium/cilium/issues/30093 and with the latest 6.1 image in ci-verifier :grimacing::

https://github.com/cilium/cilium/pull/30453

Reply:
Thanks for digging into this, Tom.

Reply:
> I think the key point of this is, the very non-linear relationship between some of these code changes and the number of instructions, coupled with the dependency on clang versions for reproduction, also suggests to me that something strange is going on with the compiler.

FWIW I don't think that this particular aspect is particularly unusual or noteworthy. The verifier doesn't operate by just iterating each instruction and determining that each of the instructions are safe and then give approval to load the BPF program. Rather, the verifier iterates through instructions, building up the expected state of the registers, stack etc. up until any fork in the control flow. Then it stores the state and starts exploring each fork, collecting state as it goes, and potentially forking out from there. Depending on the way that these forks diverge and converge, the same instruction could be revisited multiple times. There are some techniques that it implements for 'state pruning' where it can recognize when the paths converge again, and then rather than iterating through the entire state from that point forward for each of the earlier forks, it iterates through the path the first time, stores the state & validity of that path, then goes back to the other fork, identifies that it also reaches the same path and then avoids following down the same path of the program again. This is a bit of a loose explanation, there are likely other resources out there that explain this better. But let's say that the compiler compilation of the BPF code causes the generated code to no longer converge in the same way, or the kernel cannot identify converging paths, then that could mean that an entire fork of possibilities through to all possible endings/exits of the program could be evaluated twice (or more!). As such, it's possible for relatively small changes in the generated code to cause an exponential increase in the evaluated complexity of the programs from a kernel perspective.